service/matching/task_queue_partition_manager.go

```diff
@@ -0,0 +1,707 @@
+	ctx context.Context,
+	version string,
+	create bool,
+) (taskQueueManager, error) {
+	tqm, err := pm.getVersionedQueueNoWait(version, create)
+	if err != nil || tqm == nil {
+		return nil, err
+	}
+	if err = tqm.WaitUntilInitialized(ctx); err != nil {
+		return nil, err
+	}
+	return tqm, nil
+}
+
+// Returns taskQueuePartitionManager for a task queue. If not already cached, and create is true, tries
+// to get new range from DB and create one. This does not block for the task queue to be
+// initialized.
+func (pm *taskQueuePartitionManagerImpl) getVersionedQueueNoWait(
+	version string,
+	create bool,
+) (taskQueueManager, error) {
+	pm.versionedQueuesLock.RLock()
+	vq, ok := pm.versionedQueues[version]
+	pm.versionedQueuesLock.RUnlock()
+	if !ok {
+		if !create {
+			return nil, nil
+		}
+
+		// If it gets here, write lock and check again in case a task queue is created between the two locks
+		pm.versionedQueuesLock.Lock()
+		vq, ok = pm.versionedQueues[version]
+		if !ok {
+			var err error
+			vq, err = newTaskQueueManager(pm, newTaskQueueIDWithVersionSet(pm.taskQueueID, version))
+			if err != nil {
+				pm.versionedQueuesLock.Unlock()
+				return nil, err
+			}
+			pm.versionedQueues[version] = vq
+		}
+		pm.versionedQueuesLock.Unlock()
+
+		if !ok {
+			vq.Start()
+			pm.engine.updateTaskQueueGauge(pm, true, 1)
+		}
+	}
+	return vq, nil
+}
+
+func (pm *taskQueuePartitionManagerImpl) redirectToVersionSetQueueForPoll(
+	ctx context.Context,
+	caps *commonpb.WorkerVersionCapabilities,
+	userData *persistencespb.VersionedTaskQueueUserData,
+) (taskQueueManager, error) {
+	data := userData.GetData().GetVersioningData()
+
+	if pm.kind == enumspb.TASK_QUEUE_KIND_STICKY {
+		// In the sticky case we don't redirect, but we may kick off this worker if there's a newer one.
+		unknownBuild, err := checkVersionForStickyPoll(data, caps)
+		if err != nil {
+			return nil, err
+		}
+		if unknownBuild {
+			pm.recordUnknownBuildPoll(caps.BuildId)
+		}
+		return pm.defaultQueue, nil
+	}
+
+	primarySetId, demotedSetIds, unknownBuild, err := lookupVersionSetForPoll(data, caps)
+	if err != nil {
+		return nil, err
+	}
+	if unknownBuild {
+		pm.recordUnknownBuildPoll(caps.BuildId)
+	}
+	pm.loadDemotedSetIds(demotedSetIds)
+
+	return pm.getVersionedQueue(ctx, primarySetId, true)
+}
+
+func (pm *taskQueuePartitionManagerImpl) loadDemotedSetIds(demotedSetIds []string) {
+	// If we have demoted set ids, we need to load all task queues for them because even though
+	// no new tasks will be sent to them, they might have old tasks in the db.
+	// Also mark them alive, so that their liveness will be roughly synchronized.
+	// TODO: once we know a demoted set id has no more tasks, we can remove it from versioning data
+	for _, demotedSetId := range demotedSetIds {
+		tqm, _ := pm.getVersionedQueueNoWait(demotedSetId, true)
+		if tqm != nil {
+			tqm.MarkAlive()
+		}
+	}
+}
+
+func (pm *taskQueuePartitionManagerImpl) getDbQueueForAdd(
```

Comment:
does this function really add much? why not just call redirectToVersionSetQueueForAdd directly? yeah, you have to call GetUserData first, but it's not like it costs anything, it's just reading from memory

Reply:
ah, ok